app/table-streams/synonyms/index.ts

```diff
@@ -34,26 +34,59 @@ async function removeIndex(id: string) {
   }
 }
 
-async function putIndex(synonym: Synonym) {
+async function putIndex(synonymGroup: SynonymGroup) {
   const client = await getSearchClient()
   await client.index({
     index,
-    body: synonym,
+    id: synonymGroup.synonymId,
+    body: synonymGroup,
   })
 }
 
+async function getIndex(synonym: Synonym) {
+  const client = await getSearchClient()
+  try {
+    const group = await client.get({
+      index: 'synonym-groups',
+      id: synonym.synonymId,
+    })
+
+    return group.body._source as SynonymGroup
+  } catch (e) {
+    return null
+  }
+}
+
 export const handler = createTriggerHandler(
   async ({ eventName, dynamodb }: DynamoDBRecord) => {
-    const id = unmarshallTrigger(dynamodb!.Keys).id as string
+    if (!eventName || !dynamodb) return
     const promises = []
-
-    if (eventName === 'REMOVE') {
+    const id = unmarshallTrigger(dynamodb!.Keys).synonymId
+    const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
+    const existingGroup = await getIndex(synonym)
+    const group =
+      eventName === 'INSERT' && !existingGroup
+        ? { synonymId: synonym.synonymId, eventIds: [synonym.eventId] }
+        : existingGroup
+    if (!group) return
+    if (
+      eventName === 'REMOVE' &&
+      group.eventIds.length === 1 &&
+      group.eventIds.includes(synonym.eventId)
+    ) {
       promises.push(removeIndex(id))
+    } else if (eventName === 'REMOVE' && group.eventIds.length > 1) {
+      const modifiedGroup = {
+        synonymId: synonym.synonymId,
+        eventIds: group.eventIds.filter((id) => id !== synonym.eventId),
+      } as SynonymGroup
+      promises.push(putIndex(modifiedGroup))
     } /* (eventName === 'INSERT' || eventName === 'MODIFY') */ else {
-      const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
-      promises.push(putIndex(synonym))
+      if (!group.eventIds.includes(synonym.eventId)) {
+        group.eventIds.push(synonym.eventId)
+      }
+      promises.push(putIndex(group as SynonymGroup))
```

Comment:
As I said the other day, you shouldn't need to compute the terms to add or remove if you just query all of the terms that match the UUID of the DynamoDB record that triggered the Lambda.